### PR TITLE
Integrate the trie

### PR DESCRIPTION
**Pull request by Avi-D-coder**
This PR just sets up the trie thread and state.
We still need all the prior logic against the btree to pre-validate transactions.
That btree will be replaced by a cache db `(PublicKey -> Account)`.

We could remove pre-validated transaction requirement with an immutable `trie::SnapshotBuilder` that supported could rollback.
However, we would still need to prevent DoS with the cache db. Real hardware is very different from the zkVM, binary tries don't perform all that well on SSDs. 

In the future we will also need to warm up the `NodeHash -> TrieNode` DB right before a txn hits the trie state thread. By cashing `TrieNode`s into ram we can move slow IO off that thread.

Next Steps after this PR:
- Adding persistence with rocks
- Adding a validating thread for testing against (like a mock contract for testing)
- General testing txns (we can replicate what I did in the trie repo).

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
I'm going to merge the txn logic between the trie and btree-cache-db before this should be merged.

File: kairos-tx/schema.asn [hunk 1]
```diff
@@ -6,6 +6,7 @@ TxSchema DEFINITIONS AUTOMATIC TAGS ::= BEGIN
   PublicKey ::= OCTET STRING
   Amount ::= INTEGER (0..18446744073709551615)
   Nonce ::= INTEGER (0..18446744073709551615)
+  Epoch ::= INTEGER (0..18446744073709551615)
```
Comment (marijanp) [inline kairos-tx/schema.asn:9]: Shouldn't the epoch be added to SigningPayload?

File: kairos-server/src/state.rs [hunk 2]
```diff
@@ -1,33 +1,46 @@
...
-            batched_transfers: HashSet::new(),
-            batched_deposits: Vec::new(),
-            batched_withdrawals: Vec::new(),
+            balances,
+            batch_epoch,
+            trie_thread,
+            queued_transactions,
         }))
     }
+
```
Comment (marijanp) [inline kairos-server/src/state.rs:42]: Add docstring here

File: kairos-server/src/state.rs [hunk 3]
```diff
@@ -1,33 +1,46 @@
...
-    pub batched_deposits: Vec<Deposit>,
-    pub batched_withdrawals: Vec<Withdrawal>,
+
+    pub trie_thread: JoinHandle<()>,
+    pub queued_transactions: mpsc::Sender<TrieStateThreadMsg>,
 }
+
 impl BatchState {
-    pub fn new() -> Arc<RwLock<Self>> {
+    pub fn new(
```
Comment (marijanp) [inline kairos-server/src/state.rs:26]: Could you document what new does

File: kairos-server/src/state/trie.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+        new
+    }
+
+    /// Check if a transaction is in the batch.
+    fn contains(&self, txn: &Transaction) -> bool {
+        self.set.contains(txn)
+    }
+
+    #[allow(dead_code)]
+    fn get(&self, op_idx: usize) -> Option<&Arc<Transaction>> {
```
Comment (marijanp) [inline kairos-server/src/state/trie.rs:288]: what is `op_id`?

File: kairos-server/src/state/trie.rs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+        Ok(())
+    }
+}
+
+// We could use a self-referential struct here to avoid Arc, but it's not worth the complexity
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct BatchedTxns {
+    set: HashSet<Arc<Transaction>>,
+    ord: Vec<Arc<Transaction>>,
```
Comment (marijanp) [inline kairos-server/src/state/trie.rs:262]: Can you document what ord is or better: give it a verboser name. Ordered here seems to refer to the ordered-transactions, that were not already in the set

File: kairos-server/src/routes/withdraw.rs [hunk 6]
```diff
@@ -75,15 +83,47 @@ pub async fn withdraw_handler(
...
+            anyhow!("withdrawal amount must be greater than 0"),
+            StatusCode::BAD_REQUEST,
+        ));
+    }
+
     let state = state.read().await;
+
+    if state.batch_epoch != epoch {
+        return Err(AppErr::set_status(
+            anyhow!("Deposit epoch does not match current batch epoch."),
```
Comment (marijanp) [inline kairos-server/src/routes/withdraw.rs:122]:
```suggestion
            anyhow!("Withdrawal epoch does not match current batch epoch."),
```

File: kairos-server/src/state/trie.rs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+                        tracing::error!("failed to send commit result: {:?}", err);
+                    }
+                }
+            }
+        }
+    })
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct BatchOutput {
```
Comment (marijanp) [inline kairos-server/src/state/trie.rs:59]: Add docstring here

File: kairos-server/src/state/trie.rs [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    pub batch_epoch: u64,
+    pub new_root: TrieRoot<NodeHash>,
+    pub old_root: TrieRoot<NodeHash>,
+    pub snapshot: Snapshot<Account>,
+    pub batched_txns: BatchedTxns,
+}
+
+pub struct TrieState {
```
Comment (marijanp) [inline kairos-server/src/state/trie.rs:67]: Add docstring here

**Review by marijanp — COMMENTED** [review]
Minor things, mostly requests for additional documentation

**Review by koxu1996 — CHANGES_REQUESTED** [review]
Very solid PR :clap:. ~~One requested change is to remove epoch locking - as we discussed during last deep dive - so we avoid accepted transactions that are unexpectedly failing because of batch end/limit.~~ :speech_balloon: (This was already done, I missed it somehow.)

Also thanks for the improvement to `kairos-tx`, I will extract them to separate PR to avoid clutter in yours.

**Comment by Avi-D-coder** [comment]
> One requested change is to remove epoch locking - as we discussed during last deep dive - so we avoid accepted transactions that are unexpectedly failing because of batch end/limit.

I think I already did this in https://github.com/cspr-rad/kairos/pull/57/commits/74d04449271bfc89144bfcc5d947232391b93476.

I removed the my outdated comment about batch epoch.
Although I think we might still need an epoch for time expiring transactions.

**Comment by koxu1996** [comment]
What about TTL? It could be maximum timestamp, or maximum epoch.

**Review by marijanp — COMMENTED** [review]
Where I asked for more docs for types and functions, I was more in the hope of getting explanations of how the different datastructures are supposed to be used together, what caveats there are.

File: nixos/tests/end-to-end.nix [hunk 9]
```diff
@@ -47,32 +47,32 @@ nixosTest {
...
     # Tx Payload
-    #   nonce = 1
+    #   nonce = 0
     #   deposit:
     #     amount = 1000
     #
-    deposit_payload = "3009020101a004020203e8"
-    deposit_request = { "public_key": "deadbeef", "payload": deposit_payload, "signature": "cafebabe" }
+    deposit_payload = "3009020100a004020203e8"
+    deposit_request = { "public_key": "cafebabe", "payload": deposit_payload, "signature": "cafebabe" }
```
[the remark this answers is not in the document]
Comment (Avi-D-coder) [reply nixos/tests/end-to-end.nix:55]: Before it was `deadbeef` -> `deadbeef` -> `deadbeef`.
Now I am transferring between two accounts. 
The signatures I have just made match the public keys.

File: kairos-tx/src/asn.rs [hunk 10]
```diff
@@ -58,13 +76,62 @@ impl TryFrom<Nonce> for u64 {
...
+        }
+    }
+
+    pub fn new_deposit(amount: impl Into<Amount>) -> Self {
+        Self {
+            // deposits have no meaningful nonce
+            nonce: 0.into(),
+            body: TransactionBody::Deposit(Deposit::new(amount)),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Avi-D-coder) [reply kairos-tx/src/asn.rs:106]: I'd like to move nonce to transfer and withdraw in the future.

[the remark this answers is not in the document]
Comment (Avi-D-coder) [reply kairos-tx/src/asn.rs:106]: However some of this duplication is to make type inference nicer.

`SigningPayload::new` vs `SigningPayload::new_*`

File: kairos-server/src/state/transactions/batch_state.rs [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    pub fn execute_transaction(&mut self, txn: Signed<Transaction>) -> Result<(), AppErr> {
+        let amount = match &txn.transaction {
+            Transaction::Transfer(transfer) => transfer.amount,
+            Transaction::Deposit(deposit) => deposit.amount,
+            Transaction::Withdraw(withdraw) => withdraw.amount,
+        };
+
+        if amount == 0 {
```
[the remark this answers is not in the document]
Comment (Avi-D-coder) [reply kairos-server/src/state/transactions/batch_state.rs:33]: Made the change then noticed it causes.
`#[deny(clippy::absurd_extreme_comparisons)]`

I can make it this if you feel strongly?

```
        #[allow(clippy::absurd_extreme_comparisons)]
        if amount <= 0 {
```

**Comment by Avi-D-coder** [comment]
> Where I asked for more docs for types and functions, I was more in the hope of getting explanations of how the different datastructures are supposed to be used together, what caveats there are.

I added a bit more, but I'm not clear on what parts are confusing?
Are you looking for an overview of the trie logic or something else?

**Review by koxu1996 — APPROVED** [review]
:rocket:

File: kairos-server/src/state/transactions/batch_state.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (marijanp) [reply kairos-server/src/state/transactions/batch_state.rs:33]: As discussed when we talked the last time, we are not going to do this

File: kairos-tx/src/asn.rs [hunk 12]
```diff
@@ -235,7 +241,7 @@ mod tests {
             encoded,
             vec![
                 0b00110000, // T: 0b00 <- universal, 0b1 <- constructed, 0b10000 (16) <- SEQUENCE tag
-                0b00001001, // L: 0b0 <- short form, 0b0001001 (9) <- length
+                0b00001001, // L: 0b0 <- short form, 0b0001100 (9) <- length
```
Comment (marijanp) [inline kairos-tx/src/asn.rs:244]: Is this change still relevant?

File: kairos-tx/src/asn.rs [hunk 13]
```diff
@@ -14,6 +14,12 @@ use rasn::{Decode, Encode};
 #[rasn(delegate)]
 pub struct PublicKey(pub(crate) OctetString);
 
+impl From<&[u8]> for PublicKey {
+    fn from(value: &[u8]) -> Self {
+        PublicKey(OctetString::copy_from_slice(value))
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (koxu1996) [reply kairos-tx/src/asn.rs]: Fixed in https://github.com/cspr-rad/kairos/pull/57/commits/6874e52e34e3b4d028d34e867683eb0e35ae5410.

**Review by marijanp — APPROVED** [review]
Thank you for adding the comments

**Review by marijanp — APPROVED** [review]

**Pull request merged**